### PR TITLE
Don't register dask cache globally - but do use cache as context manager when slicing.  Add `cache` parameter to base layer.

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Following some conversations on zulip, here's a proof of principle approach that would have us move away from a global dask cache, and allow each dask-backed layer to have it's own dask cache that only gets activated in a context manager surrounding `_set_view_slice`

The key here is that the `configure_dask` function creates a closure on a dask `Cache`, which the layer can then use in
with `layer.dask_optimized_slicing(): ...`

cc @perlman, @jni

### CODE REVIEW

File: napari/utils/dask_utils.py [hunk 1]
```diff
@@ -170,23 +172,25 @@ def configure_dask(data) -> Callable[[], ContextManager[dict]]:
...
+                'For best performance with Dask arrays in napari, please upgrade Dask to v2.15.0 or later. Current version is {dask_version}',
+                deferred=True,
+                dask_version=dask.__version__,
             )
+        )
 
-        def dask_optimized_slicing():
-            with dask.config.set({"optimization.fuse.active": False}) as cfg:
-                yield cfg
+    _cache = Cache(nbytes)
```
Comment (tlambert03) [inline napari/utils/dask_utils.py]: this is the new closure-local cache, which used in the context at line 193

**Comment by tlambert03** [comment]
note: an alternative here is to _not_ create layer-local caches (i.e. keep the global `utils.dask_cache` ...) but rather than calling cache.register here:

https://github.com/napari/napari/blob/c6df8cfb4155d6314796ed4e7c5334646ed75665/napari/utils/dask_utils.py#L47

add it to the context manager here:

https://github.com/napari/napari/blob/c6df8cfb4155d6314796ed4e7c5334646ed75665/napari/utils/dask_utils.py#L184

**Comment by jni** [comment]
In principle I love this, but I still have questions:

- what is the lifetime of this cache object? Will it get cleared when I delete the layer?
- how can I control this cache from napari? For dynamic operations like napariboard, we actually want the cache off. It would be pretty awesome if we could control the cache behaviour *per layer!*

**Comment by tlambert03** [comment]
> what is the lifetime of this cache object? Will it get cleared when I delete the layer?

that would be my assumption, but we can test it.  If you grab a dask layer created by `napari.add_image` in the current PR, you'll  see the actual cache object at `layer.dask_optimized_slicing.__closure__[0].cell_contents.__closure__[0].cell_contents.cache` ... I believe that closure is probably the only reference to that cache and will likely be collected with the layer.

> how can I control this cache from napari? For dynamic operations like napariboard, we actually want the cache off. It would be pretty awesome if we could control the cache behaviour per layer!

we'd need to add a new layer parameter, or a `settings` object and or context manager (still coming, promise)

**Review by jni — APPROVED** [review]
❤️

**Comment by tlambert03** [comment]
This is ready for review.  I know @jni already approved it... but here's what's changed since then (chatted with @jni on zulip to make sure he's ok with the changes).  This PR turns dask caching behavior back _on_ by default.  However, unlike the previous implementation, this PR no longer calls `cache.register()` globally, which means it should not affect dask caching behavior for anyone using dask arrays outside of napari layers.

Instead, the cache is added to the "`dask_optimized_slicing`" context manager [here](https://github.com/napari/napari/pull/3285/files#diff-c2b4d5fbc3c8539b5e0d867f6c9139bd731afb468515f4b00b5c8fcb20de805cR154), which gets used only in the `set_view_slice` method [here](https://github.com/napari/napari/blob/main/napari/layers/base/base.py#L953).  There is a still a single global cache (which is desirable, so they can all share that space)... it's just not "always on".

because @jni mentioned that he'd like some layers to be able to opt-out of caching, there is a new `Layer.__init__` parameter called `cache`, which defaults to `True`.  by setting it to `False`, a single layer can opt out of caching.  And to generally opt-out of caching, one can still use `resize_dask_cache(0)` as before.

This PR does remove the `utils.dask_cache` variable... I don't think that would have been used by anybody, but technically it was in a public namespace, so just pointing that out.

**Comment by jni** [comment]
@tlambert03 I *still* approve of this, and was going to move to "awaiting merge", but the test failure is on MINREQ is genuine, unfortunately!

**Pull request merged**